sunpy/map/mapbase.py

```diff
@@ -1075,94 +1132,98 @@ def submap(self, range_a, range_b):
-                range_b.unit.is_equivalent(self.spatial_units.y)):
-                units = 'data'
-            elif range_a.unit.is_equivalent(u.pixel) and range_b.unit.is_equivalent(u.pixel):
-                units = 'pixels'
+        if isinstance(bottom_left, (astropy.coordinates.SkyCoord,
+                                    astropy.coordinates.BaseCoordinateFrame)):
+            if not top_right:
+                if bottom_left.shape[0] != 2:
+                    raise ValueError("If top_right is not specified bottom_left must have length two")
+                else:
+                    lon, lat = self._get_lon_lat(bottom_left)
+                    top_right = u.Quantity([lon[1], lat[1]])
+                    bottom_left = u.Quantity([lon[0], lat[0]])
             else:
-                raise u.UnitsError("range_a and range_b but be "
-                                   "in units convertable to {} or {}".format(self.spatial_units['x'],
-                                                                             u.pixel))
-        else:
-            raise TypeError("Arguments range_a and range_b to function submap "
-                            "have an invalid unit attribute "
-                            "You may want to pass in an astropy Quantity instead.")
-
-        if units is "data":
-            # Check edges (e.g. [:512,..] or [:,...])
-            if range_a[0] is None:
-                range_a[0] = self.xrange[0]
-            if range_a[1] is None:
-                range_a[1] = self.xrange[1]
-            if range_b[0] is None:
-                range_b[0] = self.yrange[0]
-            if range_b[1] is None:
-                range_b[1] = self.yrange[1]
-
-            x1, y1 = np.ceil(u.Quantity(self.data_to_pixel(range_a[0], range_b[0]))).value
-            x2, y2 = np.floor(u.Quantity(self.data_to_pixel(range_a[1], range_b[1])) + 1*u.pix).value
-
-            x_pixels = [x1, x2]
-            y_pixels = [y1, y2]
-
-        elif units is "pixels":
-            # Check edges
-            if range_a[0] is None:
-                range_a[0] = 0
-            if range_a[1] is None:
-                range_a[1] = self.data.shape[1]
-            if range_b[0] is None:
-                range_b[0] = 0
-            if range_b[1] is None:
-                range_b[1] = self.data.shape[0]
-
-            x_pixels = range_a.value
-            y_pixels = range_b.value
+                bottom_left = u.Quantity(self._get_lon_lat(bottom_left))
+                top_right = u.Quantity(self._get_lon_lat(top_right))
+
+            top_left = u.Quantity([bottom_left[0], top_right[1]])
+            bottom_right = u.Quantity([top_right[0], bottom_left[1]])
+
+            corners = u.Quantity([bottom_left, bottom_right, top_left, top_right])
+            coord = SkyCoord(corners, frame=self.coordinate_frame)
+            pixel_corners = self.data_to_pixel(coord)
+
+            x_pixels = u.Quantity([np.min(pixel_corners.x), np.max(pixel_corners.x)]).value
```

Comment:
That's simple!!! nice!! the skycords converted to pixel and take the min/max!! cool! nice job!!